### PR TITLE
auth,common: add lockless auth

### PR DESCRIPTION
**Pull request by tchaikov**
https://trello.com/c/aUSezBRH/365-stdmutex-etc-for-release-builds

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

File: src/common/lock_cond.h [hunk 1]
```diff
@@ -4,39 +4,51 @@
...
-  int Wait(LockMutex<lock_policy>&) {
-    return 0;
-  }
-  int Signal() {
-    return 0;
+  void wait(std::unique_lock<LockMutex<lock_policy>>&) {}
+  template<class Predicate>
+  void wait(std::unique_lock<LockMutex<lock_policy>>& lock, Predicate pred) {
+    while (!pred()) {
+      // TODO: PAUSE on x86
```
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/common/lock_cond.h]: i didn't think about this seriously while writing the "TODO" line. just felt embarrassed when looking at it.

currently the consumer of lockfree `condition_variable` would be `shared_cache`. and the semantic of `seastar::weak_ptr` is quite different than that of `std::weak_ptr`: one cannot promote a `weak_ptr` to a `shared_ptr` by `lock()`.

i will re-refactor the shared_cache to port it to the seastar properly.

**Comment by tchaikov** [comment]
@cbodley after a second thought, i think i'd be easier to address the locks in `src/auth` first in this PR. and then leave the refactory of `SharedLRU` for another PR. my short term plan is to have a working lockfree MonClient.

**Comment by tchaikov** [comment]
changelog

- added the debug version of `condition_variable`, `shared_mutex` following the model of `mutex_debug`
- do not define `LockCond<LockPolicy::SINGLE>`, as the lockfree `condition_variable` should not use the interface of `std::condition_variable`. and in the case of `SharedLRU`, if an entry is tracked by `weak_refs`, we should be able to promote the `weak_ref` to a `shared_ptr`. because, unlike in the alien world, the seastar fiber, on the shard where we manage the LRU cache, won't be racing with other fibers which could remove an entry from `weak_refs`. in other words, the predicator functor passed to `condition_variable` in `SharedLRU` never returns `false`. i will leave the change for "lockfree `condition_variable`" for another PR though.
- enable `NDEBUG` build. currently if `CMAKE_BUILD_TYPE` belongs to `Release`, `MinSizeRel` and `RelWithDebInfo`,  `-DNDEBUG` will be added to to CXXFLAGS and CFLAGS. and some classes under `src/auth` will use `std::mutex` and friends if `-DNDEBUG`. we apply the same machinery in other parts of Ceph once this change gets merged.

**Comment by tchaikov** [comment]
retest this please.

**Comment by cbodley** [comment]
cc @adamemerson

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2018-08-17_15:09:18-rados-wip-kefu-testing-2018-08-17-2127-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2018-08-18_16:46:03-rados-wip-kefu-testing-2018-08-18-2315-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2018-08-19_13:56:54-rados-wip-kefu-testing-2018-08-19-1812-distro-basic-mira/

the failure is tracked by http://tracker.ceph.com/issues/24485

**Comment by tchaikov** [comment]
retest this please.

File: src/auth/AuthClientHandler.h [hunk 2]
```diff
@@ -31,18 +33,19 @@ class AuthClientHandler {
   uint32_t want;
   uint32_t have;
   uint32_t need;
-  RWLock lock;
+  mutable SharedMutexT<lock_policy> lock;
```
Comment (cbodley) [inline src/auth/AuthClientHandler.h]: i wonder whether `AuthClientHandler` itself even needs explicit locking - doesn't `MonClient` already hold its `monc_lock` over any calls to `AuthClientHandler`? are there any users outside of `MonClient` that i'm missing?

Comment (tchaikov) [reply src/auth/AuthClientHandler.h]: agreed! the `RWLock` in `AuthClientHandler` was introduced in #1636. back to then, we were accessing `MonClient::auth` directly, see https://github.com/ceph/ceph/commit/a2eb6ae3fb57b09efdd4d7baac6871ca8dd8e79f#diff-fa6c2eba8356ae1442d1bf749beacfdfL6209 . and in #11128, we added a helper method offering accessing to `MonClient::auth`. this method guards the access to `auth` using `monc_lock` .

as all accesses to `MonClient::auth` are now protected by `monc_lock`, i will drop the `RWLock` in `AuthClientHandler` and its children classes in a follow-up PR.

probably we could differentiate the read/write access with RWLock. as we only update the auth's internal status/keyring when handling `MAuthReply`. most of them, we just call `auth->build_authorizer(service_id)`. but let's leave it to yet another follow-up PR.

**Pull request merged**